WalletWasabi/WabiSabi/Models/Decomposition/PossibleDecompositions.cs

```diff
@@ -0,0 +1,194 @@
+			{
+				var prunedSingletons = new CombinationsOfASize(orderedDenoms, maximumTotalValue, minimumTotalValue);
+				StratifiedDecompositions = ImmutableArray.Create<CombinationsOfASize>(prunedSingletons);
+			}
+			else
+			{
+				var bySize = ImmutableArray.CreateBuilder<CombinationsOfASize>(maxOutputs);
+
+				// Generate the base decompositions, one for each possible value,
+				// without pruning by minimum total value.
+				var unprunedSingletons = new CombinationsOfASize(orderedDenoms, maximumTotalValue, 0);
+				bySize.Add(unprunedSingletons);
+
+				// Extend to create combinations smaller than maxOutputs.
+				// There is still no pruning by minimum value as these
+				// decompositions are not yet complete.
+				while (bySize.Capacity - bySize.Count > 1)
+				{
+					bySize.Add(bySize[^1].Extend(maximumTotalValue, 0));
+				}
+
+				// The final extension can make use of the minimum value bound.
+				bySize.Add(bySize[^1].Extend(maximumTotalValue, minimumTotalValue));
+
+				StratifiedDecompositions = bySize.MoveToImmutable();
+			}
+		}
+
+		// Decompositions are kept separated by the size of the combination.
+		// Stratifying by size ensures that we can simultaneously keep them
+		// ordered by total value.
+		private ImmutableArray<CombinationsOfASize> StratifiedDecompositions { get; }
+
+		private Money MaximumTotalValue { get; }
+
+		private Money MinimumTotalValue { get; }
+
+		private int MaxOutputs => StratifiedDecompositions.Length;
+
+		// The final public API should only allow 2, later 3 access patterns
+		// efficiently:
+		// - When optimizing the decomposition of a specific balance, enumerate
+		//   from a specific effective cost downwards up to a limit.
+		//   Decompositions are then re-ordered after evaluating the cost
+		//   function. Below 4.0 btc maxSize can be 5 with negligible losses.
+		//   Above this amount 6 or more outputs may be required, and a
+		//   non-standard value can be forced, but generally such input balances
+		//   should just be avoided.
+		// - When evaluating likely denominations given others prevouts and
+		//   ownership proofs, we start enumerate downwards from multiple effective
+		//   values, one for each input or sum/difference of a pair of inputs,
+		//   and tally the denominations used by the closest decompositions.
+		//   Max effective cost = 2* largest input, can use round parameters to
+		//   precompute before learning prevout amounts. Max count can be hard
+		//   coded to 4 to ensure all clients can compute the same
+		//   decompositions. Likelyhood should take loss into account.
+		// - When evaluating input combinations, we need to query potentially
+		//   very large number of different balances, looking for small
+		//   decompositions with minimal losses (tight bounds). maxcount = 4? 5?
+		public IEnumerable<Decomposition> GetByTotalValue(
+			Money? maximumEffectiveCost = null,
+			Money? minimumTotalValue = null,
+			Money? minimumValue = null, // dust threshold
```

Comment:
This is not okay because optional parameters have to be the last ones in the parameters list. I would prefer an overloaded method that calls other with all the default values.

Reply:
FWIW all parameters are optional right now, I tried to refactor this in various ways but just ended up making it uglier. I would like to get back to this later, as this is not behavior changing, and things will be clearer once this code is used by more than just the strawman amount decomposition, at that point we should be able to make 2 or 3 overload methods that have some mandatory and some optional parameters.